### ISSUE TITLE
Fix deviation from the Pytorch loss function/module APIs

### ISSUE THREAD

**Issue by NiklasGustafsson** [dotnet/TorchSharp#558]
Currently, loss functions are implemented in Pytorch as either a Module, with the regular API patterns, or as a nn.functional API. TorchSharp implements a third API, which relies on delegates instead of Modules.

For example, the implementation of L1Loss looks like this:
```C#
[DllImport("LibTorchSharp")]
private static extern IntPtr THSNN_l1_loss(IntPtr srct, IntPtr trgt, long reduction);

/// <summary>
/// Function that takes the mean element-wise absolute value difference.
/// </summary>
/// <param name="reduction">Specifies the reduction to apply to the output</param>
/// <returns></returns>
public static Loss l1_loss(Reduction reduction = Reduction.Mean)
{
    return (Tensor src, Tensor target) => {
        var res = THSNN_l1_loss(src.Handle, target.Handle, (long)reduction);
        if (res == IntPtr.Zero) { torch.CheckForErrors(); }
        return new Tensor(res);
    };
}
```

This actually works pretty well, and the result can be used in the way that loss modules are in Python, except it's **not** a Module subclass and cannot participate in things that Modules can, such as save/restore the state_dict, which is a problem for some loss definitions -- those with a 'weight' parameter.

Further, the current implementation is found in torch.nn.functional, where the corresponding API is different:

https://pytorch.org/docs/stable/generated/torch.nn.functional.l1_loss.html?highlight=l1_loss#torch.nn.functional.l1_loss

I propose to fix this and align with the Pytorch API. That will be a breaking change.

@lostmsu, @oluwandabira, @nietras, @dsyme -- any thoughts?

**Comment by kaiidams** [comment]
No problem for me. This is a good change!

**Comment by alexkitasov** [comment]
better sooner than later anyway if inevitable in the context
